src/block/Hopper.php

```diff
@@ -96,8 +103,238 @@ public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player
 	}
 
 	public function onScheduledUpdate() : void{
-		//TODO
+		$this->position->getWorld()->scheduleDelayedBlockUpdate($this->position, 1);
+
+		$tile = $this->position->getWorld()->getTile($this->position);
+		if(!$tile instanceof TileHopper){
+			return;
+		}
+
+		$transferCooldown = $tile->getTransferCooldown();
+		if($transferCooldown > 0){
+			$transferCooldown--;
+			$tile->setTransferCooldown($transferCooldown);
+		}
+
+		if($this->isPowered()){
+			// If a hopper is powered it is deactivated and won't push, pull or pick up items.
+			return;
+		}
+
+		if($transferCooldown > 0){
+			// Skipping this tick because the hopper is still on cooldown of 8 ticks.
+			return;
+		}
+
+		$success = $this->push($tile);
+		// Hoppers that have a container above them, won't try to pick up items.
+		$origin = $this->position->getWorld()->getTile($this->position->getSide(Facing::UP));
+		if($origin instanceof Container){
```

Comment:
This could be simplified to:
`$success |= $origin instanceof Container ? $this->pull($tile, $origin) : $this->pickup($tile);`

Reply:
I especially didn't choose the ternary operator here because in the future, it can't be used without making everything crowded. 
Right now, it is easy to just use the Container interface as data type. But if Composters and Jukebox (Tiles that aren't Containers but still affected by hoppers) are fully implemented in the future, this if statement needs to change